### PR TITLE
🌱 clusterctl: implement CRD name precheck

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
WIP:
* unit tests, missing just looking for early feedback

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #5686

### CODE REVIEW

**Review by fabriziopandini — COMMENTED** [review]
overall lgtm
please add a unit test before removing WIP

File: cmd/clusterctl/client/cluster/installer.go [hunk 1]
```diff
@@ -165,6 +173,12 @@ func waitDeploymentReady(deployment unstructured.Unstructured, timeout time.Dura
 	})
 }
 
+// crdNameAllowList contains names of CRDs that we are aware of that we don't have to validate
+// as they are not referenced in core Cluster API CRDs.
+var crdNameAllowList = sets.NewString(
+	"metal3datas.infrastructure.cluster.x-k8s.io",
+)
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/installer.go:180]: Fine for me. Just wanted to make it easier for them. I would have dropped this ideally with v1.4 directly so they now have to add this annotation only for v1.3 which is a bit annoying

[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/installer.go:180]: Hm we can. No real objections to leaving it there in the future. The only problem is that we don't really care about CRDs which are not referenced in core ClusterAPI CRDs so if we have findings there's a high chance those are false positives.

For the ones which are relevant CAPI won't work anymore in the future (in the sense of even in a quickstart use case nothing will work).

But as I said. I'm fine with keeping it if you prefer. I'll drop the allow list

[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/installer.go:180]: cc @furkatgofurov7

File: cmd/clusterctl/client/cluster/installer_test.go [hunk 2]
```diff
@@ -257,6 +258,49 @@ func Test_providerInstaller_Validate(t *testing.T) {
...
+		{
+			name:    "install core with valid CRD",
+			crd:     newFakeCRD("clusterclasses.cluster.x-k8s.io", nil),
+			gk:      &schema.GroupKind{Group: "cluster.x-k8s.io", Kind: "ClusterClass"},
+			wantErr: false,
+		},
+		{
+			name:    "install core + infra with invalid CRD",
+			crd:     newFakeCRD("clusterclass.cluster.x-k8s.io", nil),
+			gk:      &schema.GroupKind{Group: "cluster.x-k8s.io", Kind: "ClusterClass"},
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/installer_test.go:277]: Fixed the names. Should clarify the difference :)

1.st test case has a valid name 2nd one has an invalid name

File: cmd/clusterctl/client/cluster/installer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply cmd/clusterctl/client/cluster/installer.go]: Thanks for the ping @sbueringer, I'll push a PR to the provider.

[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/installer.go]: Thank you!

**Comment by sbueringer** [comment]
@fabriziopandini I think this should be ready for merge (so we can get it into beta 1 already)

File: cmd/clusterctl/client/cluster/installer.go [hunk 3]
```diff
@@ -209,9 +216,75 @@ func (i *providerInstaller) Validate() error {
...
+
+	// Return if name is correct.
+	if obj.GetName() == correctCRDName {
+		return nil
+	}
+
+	return errors.Errorf("WARNING: CRD name %q is invalid for a CRD referenced in a core Cluster API CRD,"+
+		"it should be %q. Support for CRDs that are referenced in core Cluster API resources with invalid names will be "+
+		"dropped in a future Cluster API release. Note: Please check if this CRD is actually referenced in core Cluster API "+
+		"CRDs. If not, this warning can be hidden by setting the 'clusterctl.cluster.x-k8s.io/ignore-invalid-crd-name' annotation.", obj.GetName(), correctCRDName)
```
Comment (furkatgofurov7) [inline cmd/clusterctl/client/cluster/installer.go:285]: @sbueringer this looks to be left out after pushing the [suggestion](https://github.com/kubernetes-sigs/cluster-api/pull/7506#discussion_r1015468779)?

**Comment by sbueringer** [comment]
/hold 
for office hours

File: cmd/clusterctl/client/cluster/installer.go (hunk 3 above)
Comment (furkatgofurov7) [reply cmd/clusterctl/client/cluster/installer.go]: Thanks, LGTM.

**Comment by fabriziopandini** [comment]
/lgtm
/approve
(there is still the hold for office hours)

**Comment by elmiko** [comment]
as discussed at the community meeting on 11 November, please add your comments here if you have objections to this change.

if no objections are added by 11 November, we will remove the hold and merge this.

File: cmd/clusterctl/client/cluster/installer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply cmd/clusterctl/client/cluster/installer.go]: @sbueringer FYI https://github.com/metal3-io/cluster-api-provider-metal3/pull/783 PTAL, thanks!

**Pull request merged**